controllers/watcher_controller.go

```diff
@@ -76,28 +78,36 @@ func (r *WatcherReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ct
 	if !watcherObj.DeletionTimestamp.IsZero() {
 		err := r.IstioClient.RemoveVirtualServiceConfigForCR(ctx, req.NamespacedName)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
 		controllerutil.RemoveFinalizer(watcherObj, watcherFinalizer)
 		err = r.Update(ctx, watcherObj)
 		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+			return ctrl.Result{}, err
 		}
-		return ctrl.Result{}, nil
+		return ctrl.Result{Requeue: true}, nil
 	}
 
 	// check finalizer on native object
 	if !controllerutil.ContainsFinalizer(watcherObj, watcherFinalizer) {
-		controllerutil.AddFinalizer(watcherObj, watcherFinalizer)
-		err := r.Update(ctx, watcherObj)
-		if err != nil {
-			return ctrl.Result{Requeue: true}, err
+		if controllerutil.AddFinalizer(watcherObj, watcherFinalizer) {
+			return ctrl.Result{}, r.Update(ctx, watcherObj)
+		}
+	}
+
+	virtualService, err := r.IstioClient.GetVirtualService(ctx)
+	if apierrors.IsNotFound(err) {
+		if _, err := r.IstioClient.CreateVirtualService(ctx, watcherObj); err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to create virtual service %w", err)
 		}
-		return ctrl.Result{}, nil
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
 	}
-	err := r.IstioClient.UpdateVirtualServiceConfig(ctx, watcherObj)
+	err = r.IstioClient.UpdateVirtualServiceConfig(ctx, watcherObj, virtualService)
```

Comment:
The controller should bring the cluster to the desired state.
If the Vs does not exist, we create with the desired state.
But if it exists, we end up here. The `UpdateVirtualServiceConfig()` method does NOT configure the Vs to the desired state in terms of gateways, hosts, etc. As far as I can see it only updates `spec.http` attribute.

Reply:
Yes, good catch! the update gateway is missing here.